lib/diagnostics_channel.js

```diff
@@ -0,0 +1,141 @@
+'use strict';
+
+const {
+  ObjectDefineProperty,
+  ObjectSetPrototypeOf,
+  ObjectValues,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+// TODO(qard): should there be a C++ publisher interface?
+class ActivePublisher {
+  subscribe(subscription) {
+    if (typeof subscription === 'function') {
+      this._subscribers.push(subscription);
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
```

Comment:
Is there concrete usages that are not based upon matching the name with a regexp? And/or if there are should regexp be special cased for ease of usability (and avoiding prototype mutation of regexp)

Reply:
Could also do a match like `list.includes(channelName)`. Also, the performance of regexp is _not great_ so I decided to remove it. My [initial prototype](https://gist.github.com/Qard/b21bb8cb1be03274f4a8c585c659adf2) was actually based on regexp but I opted for a generic function-based way to match channels instead. One could implement regexp matching externally if they really _wanted_ to.

Reply:
My main concern is about people using things like the `list.includes(channelName)` and not being robust against mutation. I was merely stating the function could accept an overloaded signature that special cases `RegExp` not to remove `function` as an overload.

Reply:
It's possible. That's what it did before, but I chose to remove it as it added a bunch of complexity and I wasn't convinced it was needed in the majority of cases. Generally you'd just want to listen to specific named channels. I'm open to being convinced otherwise though. As explained at the start of the PR, this is very much an experiment. I chose to open it _now_ to gather feedback and help steer it in a direction the community is going to be happy with, and not just what _I_ think it should look like. 😅

Reply:
As the data passed to the listener is not following any specific structure it's most likely needed to subscribe with different listeners to different channels to ease interpretation of the data.
Maybe we should remove the possibility to subscribe one anything else then a single named channel per call for the first iteration?

Reply:
I want to be able to request a bunch of specific channels to all get sent as generic data to another process for processing out of band which means I _want_ them all to go to the same listener so I can handle the routing in the other process. That's why the message handler includes the channel name.

Reply:
Should we allow symbols in addition to strings? Publishers could create and publish their symbol(s) which ensures isolation and is additionally a checkpoint for consumers to test if the version of the module supports publishing via dc.

If only strings are used we should at least recommend some sort of namespacing in doc as the channel name must be a guid.

Reply:
> Maybe we should remove the possibility to subscribe one anything else then a single named channel per call for the first iteration?

I think this would be a good change.  First for the sake of complexity removal, second because you can achieve the same thing with many subscribe calls and the same function reference, right?

```javascript
function handler() {}
dc.subscribe('foo' handler)
dc.subscribe('bar' handler)
dc.subscribe('baz' handler)
```

Unless maybe there is negative performance impact of this at a certain point?  Also, as @Flarna says, it could be in the first iteration but added later if there was need.

> Should we allow symbols in addition to strings?

Yes! So many benefits to this for folks wanting to protect their code from others on the shared channel. If you did that, it would also effect the decisions around the regex/pattern matching discussion, right?

Reply:
I made some changes and added a test to verify symbol-named channels will work.

As for simplifying subscribers, we currently have several different forms of subscription:

- string match (strict equal comparison)
- symbol match (strict equal comparison)
- function match (matches by boolean return of function accepting channel name)
- regexp match (matches by result of passing channel name to regexp test method)
- catch-all (matches everything, this is used when matcher is absent)

The first two (string and symbol) can be used with `dc.channel(name)` to get the channel object directly and are therefore the _preferred_ forms, in my opinion. The three other forms are my attempts at providing an API to subscribe to multiple channels efficiently. The catch-all is the simplest as it bypasses matching altogether, but it could be somewhat expensive to listen to _everything_ therefore it may not be desirable. Function and RegExp matching have about equal code complexity but RegExp trades performance for ease of use, which I'm unsure is the right call.

We could drop the last three forms altogether and that would allow us to _only_ have channel objects. That would allow the `DiagnosticsChannel` class to be replaced with a much simpler `channel(name)` function which doesn't have to syncronize subscribers and could also drop both top-level `publish(name, message)` and `subscribe(matcher, onMessage)` functions. It's a simpler way forward, and I don't see any reason that functionality couldn't be added back later, if there was demand for it.

Another design I had considered is a method something like an `on(name, handler)` event listener, but rather than only listening to _future_ events it would also run the handler on all the _existing_ channels first before storing it to run any time a channel is created in the future. That'd leave it up to the user to decide what to do with the channel, if they want to match on the name and attach a subscriber conditionally they can do that, though it would also leave it up to them to handle unsubscribing correctly. This would make things internally simpler but somewhat more difficult to dynamically match channels. That might not be an issue though as I see dynamic matching as somewhat being an exceptional case.